### PR TITLE
Adds telecomms-manipulator circuits

### PR DESCRIPTION
**Pull request by Carbonhell**
Adds two new circuits that can interface telecomms:
1) Telecommunication interceptor, which is able to intercept radio signals and output them as variables, along with having a "no pass" var that will not let simple radio signals(from headsets) past, if true.
Inputs:
On - if true, it intercepts all signals.
No pass - if true, it stops signals from being broadcasted where possible.
Outputs:
Source - name of the speaker
Job - his job
content - his message
spans - eventual spans such as robot span(honestly i'm not too sure about this either)
frequency - the frequency of the signal, so you can compare it to get only common messages or such.
2) Text-to-radio, which is basically text-to-speech but it outputs to radio.
The no pass variable being able to disable radio signals can be disabled by any engineer by going to a subspace receiver with a multitool and clicking the red button. The receivers also list all existing text-to-speech circuits, so if someone's being a cunt with them, you can easily fuck 'em up.
:cl:
add: Adds two new circuits that interface telecomms, tcomms interceptor(input) and text-to-radio(output), which can handle and modify telecomms messages just like old NTSL, along with new stuff in subspace receivers that can disable the interceptor's behaviour of jamming signals received and list the locations of tcomms related circuits.
/:cl:

[why]: the interceptor works similar to how NTSL worked, except the possible crashes it caused. Imo it's a nice thing to have for circuit nerds like me

### CODE REVIEW

**Comment by praisenarsie** [comment]
i hope this is only for antags

**Comment by Pyko1** [comment]
Somehow I think this is a really bad idea...

When people were doping this shit with ntsl, you at least knew where to go to fix it, with this its just some circuit crap somewhere.

Should at least be a way to block this shit from telecoms.

File: hippiestation/code/game/machinery/telecomms/computers/remote_control.dm [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/obj/machinery/computer/telecomms/remote_control
+	name = "telecommunications remote control"
+	desc = "Lets the user, generally the CE, enable or disable remote control from circuitry."
+	icon_screen = "comm_monitor"
+	circuit = /obj/item/circuitboard/computer/remote_control
+
+/obj/machinery/computer/telecomms/remote_control/attack_hand(mob/user)
```
Comment (ghost) [inline hippiestation/code/game/machinery/telecomms/computers/remote_control.dm]: Please use TGUI

**Comment by Carbonhell** [comment]
@Pyko1 do you mean for the no pass thing?because you can already block it.
Or do you mean for the text-to-radio circuit?

File: hippiestation/code/game/machinery/telecomms/computers/remote_control.dm (hunk 1 above)
Comment (Carbonhell) [reply hippiestation/code/game/machinery/telecomms/computers/remote_control.dm]: is tgui really needed for a machine with a single button and nothing else?like, the other tcomms computer don't use tgui,it feels like a waste of time

**Comment by ghost** [comment]
NTSL was one of the things that made SS13.
Means to mess with stuff in amusing manners.

In fact, it's even harder to do it since it's gated behind circuits and it's not something silly cones have direct access to while the CE access restriction is still imposed.

**Comment by Carbonhell** [comment]
do not merge, this needs some balance tweaking

**Comment by Carbonhell** [comment]
Alright so, after thinking twice, it may be better to merge the functions of the remote_access computer inside the receivers, along with adding the location of all broadcasting circuits. So engineers have a way to fight malicious circuits, by locating them and disabling them.(this way i won't have to fuck with that horrible mess called tgui,too). And this would also fix a certain balance issue with the computer being the only way to disable the interceptor no-pass function. Gonna do it this way,yea

**Comment by Carbonhell** [comment]
Updated

**Comment by ghost** [comment]
Make it so you actually have to attach the circuit to tcomms shit

**Comment by Carbonhell** [comment]
? no

**Comment by Carbonhell** [comment]
DNM tag can be removed,PR is ready to be reviewed and +1ed

**Comment by ghost** [comment]
@Carbonhell You could have a circuit blocking every message hidden in a fucking smuggler's satchel under some random floor tile, or in maint, or, _shudder_, in the incinerator

**Comment by Carbonhell** [comment]
Pls review

**Comment by ghost** [comment]
It seems cool.
If, balance wise, things get fucked because of circuits being undetectable, we can revert it and tweak it more.
+1, because I think tcomms fuckery was always the best. I miss it.

**Comment by ghost** [comment]
So how far does this reach? Is this area based? Range based?

**Comment by Carbonhell** [comment]
Interceptor's range is everywhere for intercepting signals, whereas the jammer function only works on the station zlevels and not in space areas. Its coordinates + area also get listed in the receiver if the jammer is on..
The text2radio circuit always gets listed in the receiver, and its range is equal to a normal station bounced radio range(the green thing in o2 lockers)

**Comment by Carbonhell** [comment]
Code is ready eh,just needs a review and a +1

File: hippiestation/code/modules/integrated_electronics/subtypes/input.dm [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		)
+	power_draw_idle = 0
+	spawn_flags = IC_SPAWN_RESEARCH
+	var/obj/machinery/telecomms/receiver/circuit/receiver
+	var/list/freq_blacklist = list(FREQ_CENTCOM,FREQ_SYNDICATE,FREQ_CTF_RED,FREQ_CTF_BLUE)
+
+/obj/item/integrated_circuit/input/tcomm_interceptor/Initialize()
+	..()
+	receiver = new(src)
+	receiver.holder = src
```
Comment (JohnGinnane) [inline hippiestation/code/modules/integrated_electronics/subtypes/input.dm:47]: Should you put `src` into `GLOB.ic_jammers?`

Comment (Carbonhell) [reply hippiestation/code/modules/integrated_electronics/subtypes/input.dm:47]: No, it gets added to the jammer list only if its currently jamming the signal,aka if no_pass is true

File: hippiestation/code/modules/integrated_electronics/subtypes/input.dm [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		set_pin_data(IC_OUTPUT, 1, signal.data["name"])
+		set_pin_data(IC_OUTPUT, 2, signal.data["job"])
+		set_pin_data(IC_OUTPUT, 3, signal.data["message"])
+		set_pin_data(IC_OUTPUT, 4, signal.data["spans"])
+		set_pin_data(IC_OUTPUT, 5, signal.frequency)
+		push_data()
+		activate_pin(1)
+
+/obj/item/integrated_circuit/input/tcomm_interceptor/on_data_written()
+	if(get_pin_data(IC_INPUT, 2))
```
Comment (JohnGinnane) [inline hippiestation/code/modules/integrated_electronics/subtypes/input.dm:67]: Shouldn't you check if it's on as well?

Comment (Carbonhell) [reply hippiestation/code/modules/integrated_electronics/subtypes/input.dm:67]: I am,two lines after. But keep in mind that on is for intercepting the signal only, no pass can be true and on can be false to just jam the signal

File: hippiestation/code/modules/integrated_electronics/subtypes/input.dm [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		GLOB.ic_jammers |= src
```
Comment (JohnGinnane) [inline hippiestation/code/modules/integrated_electronics/subtypes/input.dm:68]: I think use the LAZY procs for doing this? Like if `ic_jammers` isn't set up or whatever

**Review by JohnGinnane — COMMENTED** [review]
Looks p good, can we test merge pls)

**Comment by Carbonhell** [comment]
Can we get this in before next year
Ayy

**Comment by Carbonhell** [comment]
Hello darkness my old friend

**Comment by ghost** [comment]
john said it was gud so i'll just do da review

**Review by ghost — APPROVED** [review]

**Comment by Carbonhell** [comment]
https://imgur.com/a/qdhee
this can be merged

**Pull request merged**